feat(nft): introduce db cache support (non-wasm target)

laruh opened the pull request:
#900 
Current version of nft cache support includes implemented sqlite db and todos for wasm target, but it covers the general logic for both targets.

[gist with docs](https://gist.github.com/laruh/248d652d88891f3cadb09fea3a441466)

shamardy commented on mm2src/coins/nft_storage/mod.rs:
```diff
@@ -0,0 +1,159 @@
+pub trait NftStorageError: std::fmt::Debug + NotMmError + NotEqual + Send {}
+
+#[async_trait]
+pub trait NftListStorageOps {
```
It's good to add unit tests for `NftListStorageOps` for both native and wasm similar to how it was done in `utxo_block_header_storage` https://github.com/KomodoPlatform/atomicDEX-API/blob/1c261cd5a39694ae13f2699a5c4761df3468e19d/mm2src/coins/utxo/utxo_block_header_storage/mod.rs#L116 This will prevent regression if there will be any changes to DB structure or methods in the future. It will also make sure that the implementation for wasm works the same way as the sqlite one. Please consider this in the next PRs.

shamardy reviewed (commented):
Great work as always! First review iteration!

laruh commented:
UPD: I moved wasm db implementation into mod wasm like we have it in `tx_history_storage`, bcz I decided it would be more readable.
(and added tables for wasm)

laruh commented on mm2src/coins/nft.rs:
```diff
@@ -28,102 +29,339 @@ const MORALIS_FORMAT_QUERY_VALUE: &str = "decimal";
+            token_address: format!("{:#02x}", req.token_address),
+            token_id: req.token_id.to_string(),
+        })
+    }
```
oh, I forgot about ok_or_else again, thanks.
Added it for some other code blocks where we call `get_nft`

laruh commented on mm2src/coins/lp_coins.rs:
```diff
@@ -287,7 +288,11 @@ use utxo::{BlockchainNetwork, GenerateTxError, UtxoFeeDetails, UtxoTx};
+
 #[cfg(not(target_arch = "wasm32"))] pub mod z_coin;
+
+use crate::nft_storage::{CreateNftStorageError, NftStorageError};
```
Done

laruh commented on mm2src/coins/nft.rs:
```diff
@@ -28,102 +29,339 @@ const MORALIS_FORMAT_QUERY_VALUE: &str = "decimal";
 
 /// `get_nft_list` function returns list of NFTs on requested chains owned by user.
 pub async fn get_nft_list(ctx: MmArc, req: NftListReq) -> MmResult<NftList, GetNftInfoError> {
+    let storage = NftStorageBuilder::new(&ctx).build()?;
+    for chain in req.chains.iter() {
+        if !NftListStorageOps::is_initialized(&storage, chain).await? {
+            NftListStorageOps::init(&storage, chain).await?;
+        }
+    }
```
Hmmmmmm, interesting. Yeah, I dont think that it is a good idea to do it now, bcz I will need to refactor my trait methods, as right now I have identical methods like init, is_initialized in traits.

according implementation
```rust
pub(crate) async fn storage(&self) -> AccountStorageResult<&dyn AccountStorage> {
        self.storage.init().await?;
        Ok(self.storage.as_ref())
    }
```
I will probably need to impl something like this:
```rust
pub(crate) async fn storage_nft_list(&self) -> StorageResult<&dyn NftListStorageOps + NftTxHistoryStorageOps> {
        self.storage.init_nft_list().await?;
        Ok(self.storage.as_ref())
    }
```

and actually refactor `update_nft`.

shamardy commented on mm2src/coins/nft_storage/sql_storage.rs:
```diff
@@ -0,0 +1,885 @@
+    Ok(sql_builder)
+}
+
+fn nft_table_builder_preimage<'a>(conn: &'a Connection, table_name: &'a str) -> MmResult<SqlQuery<'a>, SqlError> {
```
Why is this needed when it's used only once and `SqlQuery::select_from` is used a lot instead of it in multiple places?

laruh replied:
hmm, I believe it was a part of my "legacy" code, where I was playing with SqlQuery first time.
I removed it, thanks!

shamardy commented on mm2src/coins/nft_storage/sql_storage.rs:
```diff
@@ -0,0 +1,885 @@
+    validate_table_name(&table_name)?;
+
+    let sql = format!(
+        "INSERT OR IGNORE INTO {} (
+            token_address, token_id, chain, amount, block_number, contract_type, details_json
```
Interested to know why `INSERT OR IGNORE` is used here instead of `INSERT OR REPLACE`. Probably it has lower cost to ignore if the entry is in the table but then I wonder why we have a value that can change such as amount as part of the table columns.

laruh replied:
I intentionally separated INSERT ans UPDATE methods so as not to confuse the logic.
and yeah, you are right its better to return error if we try to insert already existed entries.
I added INSERT INTO in `insert_nft_in_list_sql` and `insert_tx_in_history_sql`.

> I wonder why we have a value that can change such as amount as part of the table columns

Weeell, I decided to leave it in the case if we have filters for amount in the future. Even if I remove it from table, I still have to call update method to change the amount in details json field.

shamardy commented on mm2src/coins/nft_storage/sql_storage.rs:
```diff
@@ -0,0 +1,885 @@
+    Ok(sql)
+}
+
+fn insert_chain_into_scanned_blocks_sql() -> MmResult<String, SqlError> {
+    let table_name = scanned_nft_blocks_table_name();
+    validate_table_name(&table_name)?;
+    Ok(format!("INSERT OR IGNORE INTO {} (chain) VALUES (?1);", table_name))
+}
```
I believe this can be merged with `update_last_scanned_block_sql` by using `INSERT OR REPLACE`

laruh replied:
hmm yeah, I see, in my code I use `insert` function and right after this I call `update`. Its better just to have `INSERT OR REPLACE`.
Created `upsert_last_scanned_block_sql` function.

shamardy commented on mm2src/coins/nft_storage/sql_storage.rs:
```diff
@@ -0,0 +1,885 @@
+    Ok(sql)
+}
+
+fn update_nft_amount_sql<F>(chain: &Chain, table_name_creator: F) -> MmResult<String, SqlError>
```
Ooh, I see now why `INSERT OR IGNORE` is used in `insert_nft_in_list_sql`, you update the values using `UPDATE` statement. It's up to you and what you see fit if you want to merge this with `insert_nft_in_list_sql` by using `INSERT OR REPLACE` then. I guess it's better to have different functions for updates. Why are we using `IGNORE` though, shouldn't we allow correct errors to be returned if we used `insert_nft_in_list_sql` for an entry that already exists?

laruh replied:
>  I guess it's better to have different functions for updates.

Yeah, I would like to divide the logic.

> Why are we using IGNORE though, shouldn't we allow correct errors to be returned if we used insert_nft_in_list_sql for an entry that already exists?

Yeah, now I see that its better to return the error, added INSERT INTO

shamardy reviewed (commented):
Next review Iteration! Just some questions related to sql storage

laruh commented on mm2src/coins/nft_storage/sql_storage.rs:
```diff
@@ -0,0 +1,885 @@
+    let union_sql_strings = union_sql_strings?;
+    let union_sql = union_sql_strings.join(" UNION ALL ");
+    let mut final_sql_builder = SqlQuery::select_from_union_alias(conn, union_sql.as_str(), "nft_list")?;
+    final_sql_builder.order_desc("nft_list.block_number")?;
```
Yeah, you are right. Actually I did it on purpose, bcz:
- as blockchains have different block heights, most likely tokens from different blockchains will not be mixed with each other in the final list. For example polygon nfts from biggest to smallest, than eth nfts from biggest (where the biggest height is smaller than in Polygon) to smallest.
- yeah, anyway GUI is going to use just one chain type for nft list request
- I think it's better to have some ordering than not to have,

soo I think there are no right or wrong ways, I just did some pre ordering for GUI. As you can see I need to unify tables first and only then I can sort the table.

laruh commented on mm2src/coins/nft/nft_structs.rs:
```diff
@@ -293,4 +364,39 @@ pub(crate) struct NftTransferHistoryWrapper {
+#[allow(dead_code)]
+#[derive(Clone, Debug, Deserialize)]
+pub struct NftTxHistoryFilters {
+    #[serde(default)]
+    pub(crate) receive: bool,
+    #[serde(default)]
+    pub(crate) send: bool,
```
I chose these parameters because of GUI.
![image](https://github.com/KomodoPlatform/atomicDEX-API/assets/52405288/f0cec75e-7b4d-4d0a-b464-00fb80ebe2e3)
So if both are false(or true) then I just ignore it and return all transfers. 
```rust
if filters.send && !filters.receive {
            sql_builder.sql_builder().and_where_eq("status", "'Send'");
        } else if filters.receive && !filters.send {
            sql_builder.sql_builder().and_where_eq("status", "'Receive'");
        }
```

shamardy replied:
I guess we can leave it as is since this is what the GUI team wants

laruh commented on mm2src/coins/nft_storage/sql_storage.rs:
```diff
@@ -0,0 +1,885 @@
+        })
+        .collect();
+
+    let union_sql_strings = union_sql_strings?;
```
Oh, I see, I see, thanks!
done in `get_nft_list_builder_preimage` and in `get_nft_tx_builder_preimage`

laruh commented on mm2src/coins/nft_storage/sql_storage.rs:
```diff
@@ -0,0 +1,885 @@
+            sql_builder
+                .sql_builder()
+                .and_where(format!("block_timestamp <= '{}'", filters.to_date.unwrap()));
+        }
```
oh I over complicated it, Some( ) variant is better :)
Done, thanks!

ozkanonur commented on mm2src/coins/lp_coins.rs:
```diff
@@ -287,7 +288,11 @@ use utxo::{BlockchainNetwork, GenerateTxError, UtxoFeeDetails, UtxoTx};
 pub mod nft;
 use nft::nft_errors::GetNftInfoError;
 
+pub(crate) mod nft_storage;
```
Can we stack the `mod` statements together? It will be good to see all of them in a single place.

laruh replied:
right now we have only this for nft in lp_coins.rs
```rust
pub mod nft;
use nft::nft_errors::GetNftInfoError;
```

ozkanonur commented on mm2src/coins/lp_coins.rs:
```diff
@@ -1925,12 +1930,29 @@ pub enum WithdrawError {
+    DbError(String),
 }
 
 impl From<GetNftInfoError> for WithdrawError {
     fn from(e: GetNftInfoError) -> Self { WithdrawError::GetNftInfoError(e) }
 }
 
+impl<T: NftStorageError> From<T> for WithdrawError {
+    fn from(err: T) -> Self {
+        let msg = format!("{:?}", err);
+        WithdrawError::DbError(msg)
+    }
+}
+
+impl From<CreateNftStorageError> for WithdrawError {
+    fn from(e: CreateNftStorageError) -> Self {
+        match e {
+            CreateNftStorageError::Internal(err) => WithdrawError::InternalError(err),
+        }
+    }
+}
+
```
I would like to not add coin/protocol specific things in `lp_coins` and keep it in a generic way. We should even move the other coin/protocol specific things from it, which we will do in future. So adding things like this will increase the tech debt more and more.

laruh replied:
Sorry, didnt get your idea about generics. lp_coins is a root module of coins lib, it will contain some coins specific things. in another way we need to completely refactor lp_coins and make more modules.

ozkanonur replied:
What I meant is to keep `lp_coins` as the handler of `coins` crate in a generic way rather than having `nft`, `eth`, `tenderment` specifics in it. Because this makes having `eth`, `nft`, `tendermint`, etc. modules pointless. We should use the coin/protocol specifics in their own modules.

laruh replied:
Could you give a hint, do we have such refactoring issue, or we need to create the new one?

laruh replied:
Yeah yeah :D I just wanted to make a reference to your comment in refactoring issue if we have it

ozkanonur replied:
It's part of the crate refactoring task which we initialized 5-6 months ago. Before that crates were coupled even more. But still, we need more work on this task. I don't know if we have issue created for this, but I know we all aware of it which is also a blocker for other tasks such as providing komodo sdk via C ABI. As far as I remember I already mentioned coins and protocols will be seperated crates. We don't want to compile whole coins crate for exposing tendermint protocol.

laruh replied:
moved nft error implementations into nft related mods

ozkanonur commented on mm2src/db_common/src/sql_query.rs:
```diff
@@ -316,6 +316,23 @@ impl<'a> SqlQuery<'a> {
+        validate_table_name(alias)?;
+        Ok(SqlQuery {
+            conn,
+            sql_builder: SqlBuilder::select_from(format!("({}) AS {}", union_sql, alias)),
```
Vulnerable agains't sql injection attacks

laruh replied:
Sorry how this can be manipulated if I validate table and alias names?

And for instance according this function
```rust
fn get_nft_list_builder_preimage(conn: &Connection, chains: Vec<Chain>) -> MmResult<SqlQuery, SqlError> {
    let union_sql_strings = chains
        .iter()
        .map(|chain| {
            let table_name = nft_list_table_name(chain);
            validate_table_name(&table_name)?;
            let sql_builder = SqlQuery::select_from(conn, table_name.as_str())?;
            let sql_string = sql_builder.sql()?.trim_end_matches(';').to_string();
            Ok(sql_string)
        })
        .collect::<MmResult<Vec<_>, SqlError>>()?;
    let union_sql = union_sql_strings.join(" UNION ALL ");
    let mut final_sql_builder = SqlQuery::select_from_union_alias(conn, union_sql.as_str(), "nft_list")?;
    final_sql_builder.order_desc("nft_list.block_number")?;
    drop_mutability!(final_sql_builder);
    Ok(final_sql_builder)
}
```
chain name must be matched with enum variant

well seems like I can use this [functon](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/db_common/src/sql_query.rs#L35)

ozkanonur replied:
I don't know how exactly that function works, but you put 2 values inside of the SQL without prepared parameters, in case if any of them include malicious query, it will work.

laruh replied:
Ok I rechecked, I cant use `pub fn select_from_alias`, bcz I have union of tables, and validate_table_name(table)?; will fail.
Then I will build SqlQuery by myself right in `builder_preimage` functions.

ozkanonur replied:
Can you simply build prepared statement and use it instead of `SqlQuery` ? AFAIK `SqlQuery` was created manually by Sergey which was all about string building(that what makes it vulnerable to injections).

laruh replied:
added prepared statements into `get_nft_list` and `get_tx_history`

ozkanonur commented on mm2src/coins/nft_storage/mod.rs:
please move `nft_storage` to `nft/storage`

laruh replied:
Done

ozkanonur commented on mm2src/coins/nft.rs:
```diff
@@ -28,102 +29,336 @@ const MORALIS_FORMAT_QUERY_VALUE: &str = "decimal";
+        &req.url,
+    )
+    .await?;
+    let storage = NftStorageBuilder::new(&ctx).build()?;
```
do this first in the function, so previously made computing will not be wasted when this fails

laruh replied:
Well, it is more likely that `get_moralis_metadata` will fail, but its better to avoid sending moralis requests if something else can fail after.
Moved `let storage = NftStorageBuilder::new(&ctx).build()?` to the top.

ozkanonur reviewed (changes requested):
First review iteration from my side, along with the review notes I have 1 question. When there is an i/o error(e.g file permission or exceeding disk space), what happens? Do we handle that scenario?

laruh commented:
Answering questions

> When there is an i/o error(e.g file permission or exceeding disk space), what happens?

right now I just propagate `NftStorageError` and show the whole error message

```rust
impl<T: NftStorageError> From<T> for GetNftInfoError {
    fn from(err: T) -> Self {
        let msg = format!("{:?}", err);
        GetNftInfoError::DbError(msg)
    }
}
```

> Do we handle that scenario?

I didnt cover file permission or exceeding disk space errors. I mean I return the whole db error text DbError(msg).

Well, I can add matches for sqlite related errors `rusqlite::ErrorCode::PermissionDenied`, `rusqlite::ErrorCode::DiskFull`, `rusqlite::ErrorCode::SystemIOFailure`,`rusqlite::ErrorCode::CannotOpen`, `rusqlite::ErrorCode::CannotOpen`, `rusqlite::ErrorCode::ReadOnly` for `storage.add()` functions. 
However we cant be sure that we have permission problems or not enough space according wasm db errors.
We can assume that there permission or space problems based on these errors [DbTransactionError](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/mm2_db/src/indexed_db/drivers/transaction.rs#LL16C1-L16C1), [InitDbError](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/mm2_db/src/indexed_db/drivers/builder.rs#LL15C10-L15C21). It could be `NotSupported` and `OpeningError` from `InitDbError`, and `ErrorCreatingTransaction`, `ErrorUploadingItem` from `DbTransactionError`. 
We can try to [estimate](https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.StorageManager.html#method.estimate) browser storage. But not all browsers[ support it ](https://developer.mozilla.org/en-US/docs/Web/API/StorageManager/estimate#browser_compatibility).

laruh commented:
> In other history implementations we are handling these automatically(e.g, stop fetching history when disk is full, trigger history fetching when balance is changed, etc.)

Could you provide links please? so I can check how error handling works in those implementations.

> This implementation is following quite different way I gues.

NFT cache implementation update cache in db only when `update_nft` method is called, its not in the loop. When user receive nft they should update nft info (using swipe, it depends on GUI impl). when user broadcast nft transaction, GUI call `update_nft` method themselves.

Actually method like clean_nft_cache in my future plans, also we can think how we should change logic if there is a super large number of txs. May be for the very first nft cahcing we can add all nfts from moralis and save only the last 10 transactions (for instance). If user wants to scroll and see more txs we can get and cache them additionally.

ozkanonur commented:
> > In other history implementations we are handling these automatically(e.g, stop fetching history when disk is full, trigger history fetching when balance is changed, etc.)
> 
> Could you provide links please? so I can check how error handling works in those implementations.
> 
> > This implementation is following quite different way I gues.
> 
> NFT cache implementation update cache in db only when `update_nft` method is called, its not in the loop. When user receive nft they should update nft info (using swipe, it depends on GUI impl). when user broadcast nft transaction, GUI call `update_nft` method themselves.
> 
> Actually method like clean_nft_cache in my future plans, also we can think how we should change logic if there is a super large number of txs. May be for the very first nft caching we can add all nfts from moralis and save only the last 10 transactions (for instance). If user wants to scroll and see more txs we can get and cache them additionally.

What about doing this caching on the proxy gateway? Which can save the storage of the clients.

ozkanonur commented:
> > What about doing this caching on the proxy gateway? Which can save the storage of the clients.
> 
> not sure, imagine the proxy will be unreachable, users will have problems with withdrawing

If proxy is unreachable then they can't withdraw even without caching due to not being able to reach moralist. Am I missing something?

laruh commented:
> > > What about doing this caching on the proxy gateway? Which can save the storage of the clients.
> > 
> > 
> > not sure, imagine the proxy will be unreachable, users will have problems with withdrawing
> 
> If proxy is unreachable then they can't withdraw even without caching due to not being able to reach moralist. Am I missing something?

The thing is that we are trying not to send requests to moralis every time when user wants to access their nft. 
So when user is trying to withdraw nft we use info from cache, not moralis. And yes, if proxy is unreachable we even cant send moralis requests to get info from it, bcz we send moralis requests to proxy.

ozkanonur commented:
> The thing is that we are trying not to send requests to moralis every time when user wants to access their nft.

It's not going to hit to moralis when we cache them. When you cache the request on gateway, it will directly return the cached data in less than a millisecond until the expiration time we provided.

So the data will be provided from proxy cache not moralis. So caching the data is nothing to do with moralist it self.

> UPD we can see how nft feature works after release and change nft cache logic if necessary. We cant predict right now which way is the best.

I think we should follow the proxy caching idea because of the simple reasons. This implementation comes with a lot of costs we have talked above. And, if we don't like the current behaviour, we will need to revert 2k lines of code which is quite complicated when we implement different things later on top of these.

ozkanonur commented:
> Im talking about that if we cant access the cache from proxy, we even cant use moralis as plan B.

Can you clarify this a bit more? I am not sure if I understand you correctly

laruh commented:
> > Im talking about that if we cant access the cache from proxy, we even cant use moralis as plan B.
> 
> Can you clarify this a bit more? I am not sure if I understand you correctly

If proxy is unreachable, where user is going to take nft info?

Right now I dont use moralis if something went wrong with db cache but we can improve it in the future and try to send request to moralis if db cache failed during withdrawing.

ozkanonur commented:
Are we not using moralis for withdraw operations?

laruh commented:
> Are we not using moralis for withdraw operations?

No
https://github.com/KomodoPlatform/atomicDEX-API/blob/add-nft-cache-support/mm2src/coins/eth.rs#LL882C25-L882C47
https://github.com/KomodoPlatform/atomicDEX-API/blob/add-nft-cache-support/mm2src/coins/nft.rs#L684

ozkanonur commented:
Hmm, I see now. I don't know how realistic is the scenario of not being able to connect `moralis` and keep functioning the NFTs. Proxy will only be unreacable when moralis is unreachable. And when moralis is unreachable for some long time, NFTs will not functioning well anyway.

The "caching"(in fact, this isn't even cache) in mm2 becoming a huge overhead in mm2. For each protocol we are implementing at least 2k lines of code(which looks very spaghetti) just for parsing some data. I will work on this problem generally and propose a solution within this week.

laruh commented:
> Hmm, I see now. I don't know how realistic is the scenario of not being able to connect `moralis` and keep functioning the NFTs. Proxy will only be unreacable when moralis is unreachable. And when moralis is unreachable for some long time, NFTs will not functioning well anyway.
> 
> The "caching"(in fact, this isn't even cache) in mm2 becoming a huge overhead in mm2. For each protocol we are implementing at least 2k lines of code(which looks very spaghetti) just for parsing some data. I will work on this problem generally and propose a solution within this week.

Lets wait what @shamardy and @tonymorony think about it.
I dont think that caching nft on our proxy service is a good solution now bcz:
- its centralized, even if we store public info, we should avoid keeping users info on our side if its possible.
- we already cache coin tx history on client side. we should apply the same caching rules to features. it will be unpredictable for users if one feature is cached on their side, and other is cached somewhere else.
- users usually hold nfts in wallet for a long time after purchase, nfts are not as actively transferred as coins. So the probability that the client will run out of memory is extremely small. If we are already handling disk space errors for my tx history, I believe I can do this too, I just need some links to the code where we catch such errors for sqlite and IndexedDB. 
- we will have to support caching logic on proxy service too, as you can see I don't just delete all old entries in nft list table and overwrite it by taking all the information from moralis each time. the same will be for future features, we will have to support db logic for each new feature with its unique nuances on services. so we will have spaghetti but in other places.
- if you want to start delegating some feature implementations on services, we should evaluate this for the entire project, bcz right now we have something like monolith.

laruh commented:
> I didn't get what you mean with "unpredictable for users" :) The one we are doing isn't even cache. It's just storing values to the database :D

Storing data locally on client side and on remote proxy service are different things. 
In the first, the storage is user-controlled and accessible only to the user. In the second variant, we collect data about our users  and this data is not user-controlled. I mean, in the first variant, user can access the information offline, while in the second variant, if something goes wrong, user won't see the tokens at all.

And literally unpredictable, because we store coins tx history on client side.

laruh commented:
>  >   its centralized, even if we store public info, we should avoid keeping users info on our side if its possible.

> We will need to remove moralis completely then. Enabling proxy cache is not going to hurt anything extra.

Moralis doesn't know which addresses are komodo users.

ozkanonur commented:
I am not sure if you understand me correctly. The caching is nothing similar to what we do here. If you consider caching as a problem just because it's centrealised, then why we are using moralis at all? It's literally the same thing. If you send request to `asd.com/bla/bla/2` and get the data, how does caching be a problem? Caching is not changing it's availability of privacy.

laruh commented:
> I am not sure if you understand me correctly. The caching is nothing similar to what we do here. If you consider caching as a problem just because it's centralised, then why we are using moralis at all? It's literally the same thing. If you send request to `asd.com/bla/bla/2` and get the data, how does caching be a problem? Caching is not changing it's availability of privacy.

Moralis just provides their NFT API for us, we send requests and get responses.  For them we are some user which send requests to collects some info using api key.  
We aren't going to use moralis forever, block book has NFT feature support, but it doesnt return metadata info and dont provide tx history for fungible and non fungible tokens separately. As blockbook integration in our plans, probably we can utilize some how its nft feature or they will improve their nft support. Anyway once we find suitable decentralized nft api, we can start to use it.

Im talking about collecting our users information. How acceptable it is in the crypto market to collect customer data, even if the info is available on scanners. By collecting data on our servers we can identify our users. We can encrypt users data, but then how do we update the nft information if the info is encrypted, for example? If we can encrypt and decrypt it then its fine, if not then my question about collecting customers info is open.

shamardy commented on mm2src/coins/nft.rs:
```diff
@@ -332,3 +467,296 @@ fn get_tx_status(my_wallet: &str, to_address: &str) -> TransferStatus {
+    tx: NftTransferHistory,
+    my_address: &str,
+) -> MmResult<(), UpdateNftError> {
+    if let Some(contract_type) = tx.contract_type {
+        match (tx.status, contract_type) {
+            (TransferStatus::Send, ContractType::Erc721) => handle_send_erc721(storage, chain, tx).await,
+            (TransferStatus::Receive, ContractType::Erc721) => {
+                handle_receive_erc721(storage, chain, tx, url, my_address).await
+            },
+            (TransferStatus::Send, ContractType::Erc1155) => handle_send_erc1155(storage, chain, tx).await,
+            (TransferStatus::Receive, ContractType::Erc1155) => {
+                handle_receive_erc1155(storage, chain, tx, url, my_address).await
+            },
+        }
+    } else if tx.status == TransferStatus::Send {
+        // if contract_type is None, we need to handle this token as erc721, because it is a custom standard like CRYPTO PUNKS
+        // which has almost the same behaviour as erc721 (1 owner can have 1 nft).
+        handle_send_erc721(storage, chain, tx).await
+    } else {
+        handle_receive_erc721(storage, chain, tx, url, my_address).await
+    }
```
Fewer lines of code
```rust
match (tx.status, tx.contract_type) {
    (TransferStatus::Send, None | Some(ContractType::Erc721)) => handle_send_erc721(storage, chain, tx).await,
    (TransferStatus::Receive, None | Some(ContractType::Erc721)) => {
        handle_receive_erc721(storage, chain, tx, url, my_address).await
    },
    (TransferStatus::Send, Some(ContractType::Erc1155)) => handle_send_erc1155(storage, chain, tx).await,
    (TransferStatus::Receive, Some(ContractType::Erc1155)) => {
        handle_receive_erc1155(storage, chain, tx, url, my_address).await
    },
}
```

laruh replied:
yeap, thats better, done

shamardy reviewed (commented):
Final review from my side :)

borngraced commented on mm2src/coins/nft/storage/wasm/wasm_storage.rs:
```diff
@@ -0,0 +1,235 @@
+pub(crate) struct NftListTable {
+    token_address: String,
+    token_id: BeBigUint,
+    chain: String,
```
is `chain` here same as `ticker`?

laruh replied:
No, I use `chain.to_string()` here in the case if I need to get `Chain` type from string.
And actually I believe there is no so big difference between `chain.to_string()` and `chain.to_ticker()`  if we want to use it as the index for instance, bcz both are unique.

borngraced replied:
Ok thanks. but I think if we should use `ticker` if we can for the field name just to be consistent (with other `WASM` storage impls).

laruh replied:
In sqlite version everywhere I use to_string, except scanned_nft_blocks table, bcz ticker names are shorter, so it was a little easier for me to read rows in this small table (may be there I also should use `to_string(`)). NFT feature is quite different from other implementations, it relies on blockchain type most of the moments, not platform coin ticker.

laruh commented on mm2src/coins/nft/nft_errors.rs:
```diff
@@ -98,7 +107,8 @@ impl HttpStatusCode for GetNftInfoError {
             | GetNftInfoError::TokenNotFoundInWallet { .. }
-            | GetNftInfoError::DbError(_) => StatusCode::INTERNAL_SERVER_ERROR,
+            | GetNftInfoError::DbError(_)
+            | GetNftInfoError::ParseTimestampError(_) => StatusCode::INTERNAL_SERVER_ERROR,
```
hmmm,  I put it as an internal problem, because if we get response from moralis successfully
```rust
let response = send_request_to_uri(uri.as_str()).await?;
```
other things if they fail will be our internal problems.

But may be you are right, if parsing from date str failed 
```rust
let date: DateTime<Utc> = date_str.parse()?;
```
the problem is in string and this string we get from moralis.

I will move `GetNftInfoError::ParseTimestampError(_)` to this part
`GetNftInfoError::InvalidResponse(_) => StatusCode::FAILED_DEPENDENCY,`

laruh replied:
Done

laruh commented on mm2src/coins/nft.rs:
```diff
@@ -28,102 +31,338 @@ const MORALIS_FORMAT_QUERY_VALUE: &str = "decimal";
+            let last_tx_block = NftTxHistoryStorageOps::get_last_block_number(&storage, chain).await?;
+            let nft_transfers = get_moralis_nft_transfers(&ctx, chain, last_tx_block.map(|b| b + 1), &req.url).await?;
+            storage.add_txs_to_history(chain, nft_transfers).await?;
+        }
```
Thanks, done

laruh commented on mm2src/coins/nft.rs:
```diff
@@ -28,102 +31,338 @@ const MORALIS_FORMAT_QUERY_VALUE: &str = "decimal";
+                    let nfts = cache_nfts_from_moralis(&ctx, &storage, chain, &req.url).await?;
+                    update_meta_in_txs(&storage, chain, nfts).await?;
+                    update_txs_with_empty_meta(&storage, chain, &req.url).await?;
+                },
```
Thanks for the suggestion.
I removed `?` from `match NftListStorageOps::get_last_block_number(&storage, chain).await` bcz if we call `get_last_block_number` when table doesnt exist we will get `no such table` error immediately.
So we need to match each Ok(some) Ok(none) Err variant.

I also changed `if nft_block < scanned_block` to `if scanned_block < nft_block`, bcz when last scanned block bigger than the highest block in NFT LIST table it is correct.

shamardy commented:
@laruh is this ready for review again?

laruh commented:
> @laruh is this ready for review again?

Yes, sorry I forgot to change the label

shamardy commented on mm2src/coins/nft.rs:
```diff
@@ -205,29 +205,31 @@ async fn get_moralis_nft_list(ctx: &MmArc, chain: &Chain, url: &Url) -> MmResult
-                };
-                // collect NFTs from the page
-                res_list.push(nft);
+                if nft_wrapper.contract_type.is_some() {
```
Why not use if `let Some(contract_type) = nft_wrapper.contract_type`, it can even be done better the below way to avoid nesting
```rust
let contract_type = match nft_wrapper.contract_type {
    Some(contract_type) => contract_type.0,
    None => continue,
};
```

laruh replied:
Done
I found myself thinking that maybe we just have different tastes :D First time I wanted to use if let Some(). I guess I still haven't gotten rid of the habit of checking through `if`.

shamardy replied:
> I found myself thinking that maybe we just have different tastes :D

The main idea is to avoid using unwrap since the `is_some()` check can be changed carelessly in the future by anyone without noticing it in review and then the `unwrap` will lead to panicking. The other changes come down to tastes as you mentioned :) For me I like to reduce the nesting as much as possible in the code, it looks cleaner for me that way. You can check [flat coding](https://dev.to/recursivefaults/the-life-changing-magic-of-flat-code-h0m) and [guard clauses](https://betterprogramming.pub/refactoring-guard-clauses-2ceeaa1a9da) if you want to know more about why this is considered better by many devs.

shamardy commented on mm2src/coins/nft.rs:
```diff
@@ -281,35 +283,37 @@ async fn get_moralis_nft_transfers(
-                };
-                // collect NFTs transfers from the page
-                res_list.push(transfer_history);
+                if transfer_wrapper.contract_type.is_some() {
```
Same here

laruh replied:
Done

shamardy commented on mm2src/coins/nft.rs:
```diff
@@ -348,6 +352,9 @@ async fn get_moralis_metadata(
 
     let response = send_request_to_uri(uri.as_str()).await?;
     let nft_wrapper: NftWrapper = serde_json::from_str(&response.to_string())?;
+    if nft_wrapper.contract_type.is_none() {
+        return MmError::err(GetNftInfoError::ContractTypeIsNull);
+    }
```
matching is also preferred here to avoid using unwrap later.

shamardy commented on mm2src/coins/nft/nft_errors.rs:
```diff
@@ -102,9 +104,9 @@ impl HttpStatusCode for GetNftInfoError {
-            },
+            GetNftInfoError::InvalidResponse(_)
+            | GetNftInfoError::ParseTimestampError(_)
+            | GetNftInfoError::ContractTypeIsNull => StatusCode::FAILED_DEPENDENCY,
```
Is `FAILED_DEPENDENCY` the right status code to be used if `ContractTypeIsNull`, wouldn't `NOT_FOUND` or `BAD_REQUEST` be better?

laruh replied:
I was choosing between `FAILED_DEPENDENCY` and `PRECONDITION_FAILED`.
bcz we expect that contract type is not null.
I dont think that `BAD_REQUEST` is right here, as we send request if NFT was found in db (one of the examples), or we itarate through txs (while we filtered null contract types previously).

Well `NOT_FOUND` may be is better, I will change it.

laruh replied:
Done

shamardy replied:
`NOT_FOUND` seems to be the best choice :)

shamardy reviewed (dismissed):
🚀

shamardy commented:
@ozkanonur @borngraced can you please do your final reviews for this PR, it should be part of the next release.

ozkanonur commented on mm2src/common/common.rs:
```diff
@@ -1014,3 +1015,8 @@ pub fn sha256_digest(path: &PathBuf) -> Result<String, std::io::Error> {
+
+pub fn parse_rfc3339_to_timestamp(date_str: &str) -> Result<u64, ParseError> {
+    let date: DateTime<Utc> = date_str.parse()?;
+    Ok(date.timestamp() as u64)
```
i64 to u64 might cause integer overflow, consider using `try_into` or one of the `number_type_casting` macros

laruh replied:
added try_into, thanks!

ozkanonur reviewed (approved):
one non-blocker note

borngraced commented on mm2src/coins/nft/nft_errors.rs:
```diff
@@ -49,6 +74,21 @@ impl From<GetEthAddressError> for GetNftInfoError {
+
+impl<T: NftStorageError> From<T> for GetNftInfoError {
+    fn from(err: T) -> Self {
+        let msg = format!("{:?}", err);
+        GetNftInfoError::DbError(msg)
```
I think this can be used straight away no need to declare extra var
```rust 
GetNftInfoError::DbError(format!("{:?}", err))
```

laruh replied:
thanks, done

borngraced commented on mm2src/coins/nft/storage/mod.rs:
```diff
@@ -0,0 +1,176 @@
+
+impl<T: NftStorageError> From<T> for WithdrawError {
+    fn from(err: T) -> Self {
+        let msg = format!("{:?}", err);
+        WithdrawError::DbError(msg)
```
same here

laruh replied:
Done

borngraced commented on mm2src/coins/nft/storage/wasm/wasm_storage.rs:
```diff
@@ -0,0 +1,235 @@
+    transaction_hash: String,
+    chain: String,
+    block_number: u32,
+    block_timestamp: u64,
```
remember to use `BeBigUint` instead of `u64`.. I guess you know the reason why 👀

laruh replied:
:sweat_smile: yeah, I decided not to touch wasm storage, since Ive already added a lot of changes in `nft-cache-support-wasm` branch. well lets add `BeBigUint` in this branch too, anyway I have to resolve conflicts after merging this branch with native feature.

borngraced commented on mm2src/coins/nft/nft_errors.rs:
```diff
@@ -64,12 +104,98 @@ impl HttpStatusCode for GetNftInfoError {
+
+impl<T: NftStorageError> From<T> for UpdateNftError {
+    fn from(err: T) -> Self {
+        let msg = format!("{:?}", err);
+        UpdateNftError::DbError(msg)
```
same as above

laruh replied:
Done

borngraced commented on mm2src/coins/nft/storage/wasm/mod.rs:
```diff
@@ -0,0 +1,59 @@
+
+#[derive(Debug, Display)]
+pub enum WasmNftCacheError {
+    ErrorSerializing(String),
+    ErrorDeserializing(String),
+    ErrorSaving(String),
+    ErrorLoading(String),
+    ErrorClearing(String),
```
I think `Error` should the suffix instead of prefix here e.g `LoadingError`..wdyt?

laruh replied:
Yeah, I understand what you mean. I just used similar enumeration names as in `DbTransactionError`. Its a common practice in our project when we deal with IDB as I can see.
![image](https://github.com/KomodoPlatform/atomicDEX-API/assets/52405288/b7894289-69f8-41b7-b0e7-2ed5597943e0)

borngraced replied:
Ah okay cool

borngraced reviewed (changes requested):
great work! couple notes please

laruh commented on mm2src/coins/nft/nft_errors.rs:
```diff
@@ -21,15 +23,38 @@ pub enum GetNftInfoError {
+        match e {
+            SlurpError::ErrorDeserializing { .. } => GetNftInfoError::InvalidResponse(error_str),
+            SlurpError::Transport { .. } | SlurpError::Timeout { .. } => GetNftInfoError::Transport(error_str),
+            SlurpError::Internal(_) | SlurpError::InvalidRequest(_) => GetNftInfoError::Internal(error_str),
```
:thinking: well, `SlurpError` doesnt have `InvalidResponse` but yeah, we match it with `InvalidResponse` or `Transport`.
As for `SlurpError::InvalidRequest` sometimes we match it with `InvalidRequest` or with `Internal` . I think bcz we use custom `slurp_req` functions, and `slurp_req_body` can propagate `SlurpError::Internal` error.
But  yeah, I think its better to use `InvalidRequest` matching to avoid misunderstanding.

Changed matches for `GetNftInfoError` and `GetInfoFromUriError`

borngraced replied:
great! thanks
